influxdb3_catalog/src/serialize.rs

```diff
@@ -135,10 +138,12 @@ impl<'a> From<&'a TableDefinition> for TableSnapshot<'a> {
         let cols = def
             .schema()
             .iter()
-            .map(|(col_type, f)| {
+            .enumerate()
```

Comment:
I don't know if using `enumerate` to determine the column ID is a good idea. I think that generally, columns are always appended, in which case, it is okay, but in the event that we allow for dropping columns, then this would change their order and mess up the IDs.

We probably need some way to generate the IDs, based on what was the largest already used ID for a given table, and then ensure that that ID remains fixed for the column it is applied to for all time.

Reply:
Ah yes, that's a good catch. They should have a well defined ID that remains static regardless of what schema changes later happen to the table.

Reply:
Okay I've taken a look at this more @pauldix and @hiltontj and I've come to the conclusion we either use enumerate or we don't even bother with a column id. Here's why:

- [Here's where we define a new TableDefinition](https://github.com/influxdata/influxdb/blob/277a6e0a877d0b1068618bebbed61407c81e7abc/influxdb3_catalog/src/catalog.rs#L629-L635), it's what ends up in our Catalog when serialized
- When we create a new one [here is where we see ](https://github.com/influxdata/influxdb/blob/277a6e0a877d0b1068618bebbed61407c81e7abc/influxdb3_catalog/src/catalog.rs#L644)what we use to make a new Schema
- This becomes a [`Schema`](https://github.com/influxdata/influxdb3_core/blob/1eaa4ed5ea147bc24db98d9686e457c124dfd5b7/schema/src/lib.rs#L148-L155)
- Which is just a wrapper around [arrow::datatypes::SchemaRef](https://arrow.apache.org/rust/arrow/datatypes/type.SchemaRef.html)
- This has no way to add IDs for columns and the way most arrow stuff works is by indexing on a column essentially. There is no stable Id for a column.

I don't think this is viable beyond doing enumerate or we'd have to upstream changes to arrow itself and that seems like it wouldn't be worth the trade off or something that would make sense upstream. I don't think this change is worth it given how everything in arrow works off indexing or column name not id.

Reply:
We could have our own `TableSchema` which includes the `SchemaRef` and also includes a map of column name to id. Then the `TableSchema` is what we serialize in the catalog. We want to have the ids and not the string identifiers in the WALContents because that's much cheaper to serialize and deserialize. Also cheaper to index when inserting the data into the `WriteBuffer`.

So we don't need to update arrow, we just need a wrapper around the arrow struct where we can add our own stuff.